Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs

```diff
@@ -0,0 +1,263 @@
+        }
+
+        /// <summary>
+        /// Returns the Application Default Credentials. Subsequent invocations return cached value from
+        /// first invocation.
+        /// See <see cref="M:Google.Apis.Auth.OAuth2.GoogleCredential.GetApplicationDefaultAsync"/> for details.
+        /// </summary>
+        public Task<GoogleCredential> GetDefaultCredentialAsync()
+        {
+            return cachedCredentialTask.Value;
+        }
+
+        /// <summary>Creates a new default credential.</summary>
+        private async Task<GoogleCredential> CreateDefaultCredentialAsync()
+        {
+            // 1. First try the environment variable.
+            string credentialPath = GetEnvironmentVariable(CredentialEnvironmentVariable);
+            if (!String.IsNullOrWhiteSpace(credentialPath))
+            {
+                try
+                {
+                    return CreateDefaultCredentialFromFile(credentialPath);
+                }
+                catch (Exception e)
+                {
+                    // Catching generic exception type because any corrupted file could manifest in different ways including 
+                    // but not limited to the System, System.IO or from the Newtonsoft.Json namespace.
+                    throw new InvalidOperationException(String.Format("Error reading credential file from location {0}: {1}"
+                                            + "\nPlease check the value of the Environment Variable {2}",
+                                            credentialPath,
+                                            e.Message,
+                                            CredentialEnvironmentVariable));
+                }
+            }
+            
+            // 2. Then try the well known file.
+            credentialPath = GetWellKnownCredentialFilePath();
+            if (!String.IsNullOrWhiteSpace(credentialPath))
+            {
+                try
+                {
+                    return CreateDefaultCredentialFromFile(credentialPath);
+                }
+                catch (FileNotFoundException)
+                {
+                    // File is not present, eat the exception and move on to the next check.
+                    Logger.Debug(String.Format("Well-known credential file {0} not found.", credentialPath));
+                }
+                catch (Exception e)
+                {
+                    throw new InvalidOperationException(String.Format("Error reading credential file from location {0}: {1}"
+                                        + "\nPlease rerun 'gcloud auth login' to regenerate credentials file.",
+                                        credentialPath,
+                                        e.Message));
+                }
+            }
+
+            // 3. Then try the compute engine.     
+            Logger.Debug("Checking whether the application is running on ComputeEngine.");
```

Comment:
Not sure if this is an excellent explanation, but it's because the check actually sends an http request to metadata server, which is something the user might wanna know about (previous cases only look for local files which is probably less important).